File: matrix_multi_no_threads.py
# Referenced Youtube and Chatgbt for coding assistance 

import random
import time

def generate_random_matrix(rows, cols): # Multiplication
    return [[random.randint(0, 9) for _ in range(cols)] for _ in range(rows)]

def matrix_multiplication(A, B): # Multiplication 
    rows_A, cols_A = len(A), len(A[0])
    rows_B, cols_B = len(B), len(B[0])

    if cols_A != rows_B: # Multiplication for A and B
        raise ValueError("Number of columns in A must equal number of rows in B.")

    result = [[0] * cols_B for _ in range(rows_A)] # Multiplication for B and A
    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                result[i][j] += A[i][k] * B[k][j]  # Multiplication result

    return result

def generate_ones_matrix(size): # Check the correctness of the multiplication
    return [[1 for _ in range(size)] for _ in range(size)]

if __name__ == "__main__":
    size = 3  # Correctness proof
    A = generate_ones_matrix(size)
    B = generate_ones_matrix(size)

    result = matrix_multiplication(A, B) 

    print("\nResult Matrix:") # Print the result
    for row in result:
        print(row)


if __name__ == "__main__":  
    sizes = [10, 50, 100, 500, 1000, 1200, 1500] 

    for size in sizes:
        A, B = generate_random_matrix(size, size), generate_random_matrix(size, size)

        start_time = time.time()
        _ = matrix_multiplication(A, B)
        end_time = time.time()

        print(f"Python (No Threads), size={size} => {end_time - start_time:.4f} sec") # Print the result

File: matrix_multi_no_threads.c
// Referenced Youtube and Chatgbt for coding assistance 

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to generate a random matrix
int** generate_random_matrix(int size) {
    int** mat = (int**)malloc(size * sizeof(int*));
    for(int i = 0; i < size; i++) {
        mat[i] = (int*)malloc(size * sizeof(int));
        for(int j = 0; j < size; j++) {
            mat[i][j] = rand() % 10; // Random values from 0 to 9
        }
    }
    return mat;
}

// Function to free allocated memory by a matrix 
void free_matrix(int** mat, int size) {
    for(int i = 0; i < size; i++) {
        free(mat[i]);
    }
    free(mat);
}

// Print a matrix
void print_matrix(int** mat, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            printf("%d ", mat[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

// Matrix multiplication
void matrix_multiplication(int** A, int** B, int** C, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            C[i][j] = 0;
            for (int k = 0; k < size; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

int main() {
    srand(time(NULL)); // Random number generation

    // Matrix sizes to test
    int sizes[] = {3, 10, 50, 100, 500, 1000, 1200, 1500};  
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);

    for(int s = 0; s < num_sizes; s++) {
        int size = sizes[s];

        // Generate two random matrices
        int** A = generate_random_matrix(size);
        int** B = generate_random_matrix(size);
        int** C = (int**)malloc(size * sizeof(int*));
        for (int i = 0; i < size; i++) {
            C[i] = (int*)malloc(size * sizeof(int));
        }

        // Measure execution time
        clock_t start = clock();
        matrix_multiplication(A, B, C, size);
        clock_t end = clock();

        double elapsed_time = (double)(end - start) / CLOCKS_PER_SEC;
        printf("C (No Threads), size=%d => %.4f sec\n", size, elapsed_time);

        // Print matrices only for small sizes
        if (size == 3) {
            printf("\nMatrix A:\n");
            print_matrix(A, size);
            printf("Matrix B:\n");
            print_matrix(B, size);
            printf("Result Matrix C:\n");
            print_matrix(C, size);
        }

        // Free allocated memory
        free_matrix(A, size);
        free_matrix(B, size);
        free_matrix(C, size);
    }

    return 0;

    
}
